### DISCUSSION TITLE
multi-admin?

### DISCUSSION THREAD

**Discussion by javierbarellano4v** [mfucci/node-matter#202] [Q&A]
This may touch on multi-admin functionality...

Once I commission node-matter, e.g., on Alexa, if I go to Google Home or Apple Home, what's the expected behavior when I go to a different ecosystem (using my example, not Alexa).

Will node-matter "show up" or does it need to added and it will indicate it's already a commissioned matter device?

Thanks.

**Comment by Apollon77** [comment]
From what I know the idea is that you "go to the main controller" (the first one you paired it with) and get a new "pairing code" to be used by other controllers then to pair them too. On Apple Home there is an option in the settings of that device. Google and Alexa should also have that (and please note that for Alexa right now the device needs to be paired with Alexa as the first controller, else it will not work!!).

So yes, in fact we need to add a functionality to node-matter to also receive such a "get new pairing code" after the device was paired with node-matter as first controller.
Also we (at least not me) did not yet tested the "pair as secondary controller" functionality ... on my list

**Reply by Apollon77** [reply]
PS: a good guide how all work "in general", see https://www.tp-link.com/de/support/faq/3564/

**Reply by javierbarellano4v** [reply]
@Apollon77 Greatly appreciate this detailed info.... 

I did want to ask on the guide you reference from TP-Link (link).

Is Alexa prompting the user to select a WiFi  network to connect the "device" to? (assumes that the "device" is not connected to a network). Correct? Then, the user enters the access point's password/phrase?

**Reply by javierbarellano4v** [reply]
....and only specific to the TP-Link router or not?

**Reply by Apollon77** [reply]
I think the Wifi part is special for devices that are commissioned into wifi ... so not relevant for noe-matter right now ... these steps will not be shown

**Comment by JimBuzbee** [comment]
I did some testing regarding the "main controller" and node-matter. First, my install seems a bit fragile and I do get inconsistent results which may imply timing issues.  Generally, here's my setup where I have @mfucci  's latest "main" baseline patched to allow me to get the state of the physical device being managed by node-matter (old Wemo)

- _Google Home_: At the moment, I can pair, but the node-matter device always shows as offline.
- _Amazon Alexa Echo Dot (3rd Gen)_: Fully functional
- _SmartThings_: Can pair, but as mentioned previously, SmartThings always sends "on" command to node-matter, never "off"

If I start with Amazon and pair node-matter, I can then get a pairing code and use it with both Google Home and SmartThings to successfully pair. Behavior is as above. The Alexa app controls the Wemo and shows proper state if I change it externally. SmartThings can sends its normal "on" command.

If I start with Google Home, I can pair, but the device is listed as "offline". I can still get a pairing code and use it with Alexa which becomes fully functional as above.  I have not been successful using Google's pairing code with SmartThings.

If I start with SmartThings and pair node-matter, I don't see a way to get a pairing code, but it does give me the option of sharing the device with Google Home, but not with Alexa. If I tell it to share with Google Home, I can successfully pair, but as above, the device is shown as "offline"

**Reply by Apollon77** [reply]
Cool, thank you!

For google: Do you have a Google Nets Hub or such? In general see details I collected in https://github.com/mfucci/node-matter/pull/200/files

**Reply by JimBuzbee** [reply]
I have a couple of 1st-Gen Google Home Hubs on my network, but most all of my interactions is via the Android Google Home App.

**Reply by javierbarellano4v** [reply]
@JimBuzbee Great info too!

I need time digest all of this from @Apollon77 @JimBuzbee - and try out. Working on other things at the moment.

Appreciate it!

**Reply by JimBuzbee** [reply]
Here's a little bit a followup. As mentioned, my node-matter "device" is paired with Google Home, but it was shown as "Not Responding" and "Offline". I left node-matter running overnight and in the morning, the device was up and completely functional. I'm not sure what that implies, but node-matter was running on my Mac laptop which was closed overnight, so the network would have been off. Hmmm...

While it was working, I obtained a new pairing code and was able to pair it with Alexa. When both are paired, both are able control my physical light and both show the proper status even when controlling the light with its native app. Nice! I tried to bring SmartThings into the fold, but failed using a Google Home pairing code and with an Alexa pairing code. 

And when I restarted node-matter to start a clean test, I can pair with Google Home as usual, but once again, my device is shown as offline. :-(

**Reply by Apollon77** [reply]
I have comparable effects on apple. Directly after pairing it showed as offline, then I need to "open" the device and then status gets updated ... WIll also re-check on Google next days ...

I think it is something about announcing/discovery ... maybe it is also enough to just kill the google home app and restart? That also worked on apple

**Reply by JimBuzbee** [reply]
@Apollon77 - Thanks, I think the hint of killing the Google Home app and restarting it after pairing it helped. 

As far as SmartThings using a shared pairing code, I wonder it there's a pairing code timeout? The codes only appear to be valid for a few minutes and SmartThings pairing always seems to take a while.

**Reply by Apollon77** [reply]
If you like provide a log with a secondary pairing with google or such and one with smartthings.... maybe we can see a difference ...

**Reply by JimBuzbee** [reply]
It's not a timeout with SmartThings. Google's pairing code timeout is only 3(?) minutes but Alexa's is 15 and it still failed. The following appears to be the issue:

2023-0-5 12:19:14.443 INFO CaseServer Case server: Received pairing request from udp://fe80::da80:21d5:7a4:c66c%en0:5540
2023-0-5 12:19:14.446 ERROR CaseServer An error occured during the commissioning Error: Fabric cannot be found from destinationId

**Reply by Apollon77** [reply]
when exactly that happens?

**Reply by JimBuzbee** [reply]
@Apollon77 - I've enclosed a log file of the following sequence:

- Normal, successful pair with Google Home
- Successful pair with Alexa using a pairing code from Google Home App
- Attempt to pair to SmartThings using a pairing code from the Alexa App

The final attempt to pair with SmarThings fails after several minutes with a message of "Something went wrong. Try again later."  I do not see the same error in the log that I listed previously.

Interspersed in the log you'll see "curl" output from node-matter as I verified that things were working properly.

[GoogleAlexaSmartThings.txt](https://github.com/mfucci/node-matter/files/10523607/GoogleAlexaSmartThings.txt)

**Reply by mfucci** [reply]
This is kind of magical since I didn't implement anything special for multi-admin.... :-)
I guess now I understand better why getting Android / iOs pairing was so hard: they are doing a double provisioning using the flow as would another fabric provider...

**Comment by javierbarellano4v** [comment]
When attempting to get a new pairing code as recommended to pair with other controller (chip-tool in this case).

I get an error. Any clues?


off: IamOff
```
2023-1-3 11:30:45.014 DEBUG ExchangeManager sending id:9766/56348/214004381 t:1/9 acked:121306341 reqAck:true payload: 152800360115350137002400012401062402001835012400001818181824ff0118
2023-1-3 11:30:45.036 DEBUG MessageExchange onMessageReceived id:9766/56348/121306342 t:0/16 acked:214004381 reqAck:false payload: 
2023-1-3 11:30:45.156 DEBUG MessageExchange new MessageExchange 56349 300 2
2023-1-3 11:30:45.156 DEBUG MessageExchange onMessageReceived id:9766/56349/121306343 t:1/2 reqAck:true payload: 153600172402012403062404001818290324ff0118
2023-1-3 11:30:45.156 DEBUG InteractionProtocol Received read request from udp://192.168.1.27:5541 on session secure/9766: MA-onofflight(0x1)/On/Off(0x6)/onOff(0x0)
2023-1-3 11:30:45.157 DEBUG ExchangeManager sending id:9766/56349/214004382 t:1/5 acked:121306343 reqAck:true payload: 153601153501240000370124020124030624040018280218181824ff0118
2023-1-3 11:30:45.177 DEBUG MessageExchange onMessageReceived id:9766/56349/121306344 t:1/1 acked:214004382 re2023-1-3 11:35:21.243 DEBUG MessageExchange new MessageExchange 56352 300 2
2023-1-3 11:35:21.244 DEBUG MessageExchange onMessageReceived id:9766/56352/121306352 t:1/10 reqAck:true payload: 152500102724ff0118
2023-1-3 11:35:21.244 DEBUG InteractionProtocol Received timed request from udp://192.168.1.27:5541 on session secure/9766
2023-1-3 11:35:21.244 DEBUG ExchangeManager sending id:9766/56352/214004388 t:1/1 acked:121306352 reqAck:true payload: 1524000024ff0118
2023-1-3 11:35:21.257 DEBUG MessageExchange onMessageReceived id:9766/56352/121306353 t:1/8 acked:214004388 reqAck:true payload: 1528002901360215370024000024013c24020018350125008403300161d4adfcf70b99e5e98f2d176307beb3f8015828c79faf2706bee984e292f990b4046fef62c005770fae687c711a16d4a863194a8655abfe8f36a15160ee1da4f0e1774541ed8fc43840b4e4cd70dc14d54ce958c08059dd2879b72fcac573e8bcee25024f0225038813300413534d415254484f4d455f50414b455f53414c5418181824ff0118
2023-1-3 11:35:21.258 DEBUG InteractionProtocol Received invoke request from udp://192.168.1.27:5541 on session secure/9766: 0x0/0x3c/0x0
2023-1-3 11:35:21.264 DEBUG ExchangeManager sending id:9766/56352/214004389 t:1/9 acked:121306353 reqAck:true payload: 1528003601153500370024000024013c2402001835011818181824ff0118
2023-1-3 11:35:21.294 DEBUG MessageExchange onMessageReceived id:9766/56352/121306354 t:0/16 acked:214004389 reqAck:false payload: 
2023-1-3 11:35:22.087 DEBUG MessageExchange new MessageExchange 56353 300 2
2023-1-3 11:35:22.087 DEBUG MessageExchange onMessageReceived id:9766/56353/121306355 t:1/10 reqAck:true payload: 152500102724ff0118
2023-1-3 11:35:22.088 DEBUG InteractionProtocol Received timed request from udp://192.168.1.27:5541 on session secure/9766
2023-1-3 11:35:22.088 DEBUG ExchangeManager sending id:9766/56353/214004390 t:1/1 acked:121306355 reqAck:true payload: 1524000024ff0118
2023-1-3 11:35:22.389 DEBUG ExchangeManager sending id:51473/56353/214004390 t:1/1 acked:121306355 reqAck:true payload: 1524000024ff0118
2023-1-3 11:35:22.820 DEBUG MessageExchange onMessageReceived id:9766/56353/121306357 t:0/16 acked:214004390 reqAck:false payload: 
2023-1-3 11:35:22.923 DEBUG MessageExchange onMessageReceived id:9766/56353/121306356 t:1/8 acked:214004390 reqAck:true payload: 1528002901360215370024000024013c24020018350125008403300161bbbff487ea9cdb8939950543365e49a693b0f3b7194ff1662206c0644076171204479d4e7da85dc6ad54d9fdec124a25e820b2a5a8d170f09f295715f8f0fc825a7004489b15899e37e205d8c1a924e7848bd9be53000881192d9f668efee0bea72502060225038813300413534d415254484f4d455f50414b455f53414c5418181824ff0118
2023-1-3 11:35:22.923 DEBUG InteractionProtocol Received invoke request from udp://192.168.1.27:5541 on session secure/9766: 0x0/0x3c/0x0
2023-1-3 11:35:22.930 DEBUG ExchangeManager sending id:9766/56353/214004391 t:1/9 acked:121306356 reqAck:true payload: 1528003601153500370024000024013c2402001835011818181824ff0118
**2023-1-3 11:35:22.931 ERROR MdnsBroadcaster Error: send EHOSTUNREACH ff02::fb:5353**
2023-1-3 11:35:23.231 DEBUG ExchangeManager sending id:51473/56353/214004391 t:1/9 acked:121306356 reqAck:true payload: 1528003601153500370024000024013c2402001835011818181824ff0118
2023-1-3 11:35:23.702 DEBUG MessageExchange onMessageReceived id:9766/56353/121306359 t:0/16 acked:214004391 reqAck:false payload: 
```

![image](https://user-images.githubusercontent.com/97479879/216125975-66d30834-983d-4936-90d1-7b84773fb21b.png)

**Reply by javierbarellano4v** [reply]
setup: Alexa app on Android + Echo Dot + node-matter (MacOS/MacBook)

**Reply by mfucci** [reply]
The error above is a MDNS broadcast that should not prevent the pairing.
Probably this is broadcasting on a bad interface, as reported in https://github.com/mfucci/node-matter/issues/184
and should have no functional impact.

**Reply by Apollon77** [reply]
I had comparable isues with Alexa today ... My Alexa app got an update some days ago ... For me it wents a but further but to error on Alexa comes when it goes into certificate signing ... For me that looks the same "point" as with Google when you not whitelist the dev-decices ... The worst case would be if Alexa blocked dev devices wirth self signed certs now.

Can you try to get the Alexa version from end december on your device and try again??

**Comment by JimBuzbee** [comment]
My recollection is that the Mdns error you see is harmless. It just means that we tried to send over all interfaces and one wasn't reachable?  Anyway, I have the same setup as you and just did the following:

Pair node-matter starting with Google Home.
On Google Home, went into the device Settings/Linked Matter apps/Link apps & services/Use pairing code/copied
On Alexa, Add Device/Other/Matter/etc... Use pairing code

This was successfully for me. Alexa can now control the node-matter "device

**Reply by javierbarellano4v** [reply]
Everything worked fine with current setup (echo dot, Alexa app android), initial pairing and uttering light off/on just fine. Obtaining pairing code did not ..

@JimBuzbee previously in the thread I think you indicated you were able to get a new pairing code after the initial pairing with your Alexa. No? 

I will have to try Google .. just purchased Nest Mini. Thanks.

**Reply by javierbarellano4v** [reply]
@JimBuzbee I ask question just to verify for reference. Thanks again !

**Reply by JimBuzbee** [reply]
@javierbarellano4v - No problem, we're all here to learn :-)

I just tried again starting with Alexa first - Paired, got code from Alexa app, entered code into Google Home during add - Success!

**Reply by javierbarellano4v** [reply]
Great!

one last rudimentary question on the Google use-case, after you select 'set up device->New Device' on Google Home App on Android,  and you select a home, are you suppose to see 'matter-enabled device' along with the usual options (chromecast/google tv, camera. ...)?

**Reply by javierbarellano4v** [reply]
Pardon all the questions....

I running Google Home App (v2.63.1.12) on Android tablet (Android 10). Node-matter is running either on MacOS or on a Raspberry Pi 4.

Do I need to do something special (configure) on the Android tablet to be given a matter-enabled device option on device setup in the app?

**Reply by JimBuzbee** [reply]
@javierbarellano4v - For some reason, that option comes and goes probably as Google pushes updates. To get it back, use the camera to focus on the QR code. You should get an option to "set up device" and then it will do the Matter pairing. Then afterword, the option will be back at least until Google updates again. I find it quicker to always use the camera.

**Reply by mfucci** [reply]
The trick to get matter-enabled device option in Android is to look at a Matter QR code with the camera app and click on the link appearing...
Somebody gave me this trick and it is working well.

**Reply by javierbarellano4v** [reply]
@mfucci so then use the default camera app on the android tablet to scan the QR code?

**Reply by mfucci** [reply]
Yes, give it a shot!
1. open the camera app
2. look at the QR code
3. touch the QR code on image
4. a link  should appear below the image
5. clicking on the link will ask you to launch Google Home and it will update whatever needs to be updated to pair Matter devices

**Reply by javierbarellano4v** [reply]
@mfucci for some reason my camera app on my tablet doesn't pick up the QR code. I tried scanning QR code on the HSN on TV (yes, home shopping network!) with the android  camera app and that worked. Unfortunately, an issue on my android tablet... not sure why.

But on a different front, I had updated my Alexa app recently on my android tablet and when I tried to pair node-matter again (deleted it first), the app pairing flow changed and don't get the option to scan code!! Was hoping someone can confirm.

See sequence. I clicked to confirm that my device has a Matter logo... hoping to get Qr code scan but instead asks to see if Echo device i turning orange:

![image](https://user-images.githubusercontent.com/97479879/216388796-9fecf04c-7098-4aae-af8b-e16d95f6eee3.png)

![image](https://user-images.githubusercontent.com/97479879/216396744-a8026885-2a5a-4727-960a-fff9baed5b52.png)

![image](https://user-images.githubusercontent.com/97479879/216389917-1df71e57-8587-4085-b044-404697ce509c.png)

**Reply by javierbarellano4v** [reply]
Back to Google Home use case, I installed a QR code scanner which launched google home and the option for a 'matter-enabled device' now shows. Unfortunately, during the setup after showing Matter logo and connecting, it eventually errors out in the Google Home App. I see debug info displaying with no errors on node-matter side.

**Reply by Apollon77** [reply]
For me in Google home also the Matter option disapeared :-( SO I think Ined to get a QR scanner too ...

**Reply by Apollon77** [reply]
Ok, cool ... I just needed to try 3 QR scanner apps for it to work :-)

**Comment by JimBuzbee** [comment]
@javierbarellano4v If you haven't created a "project" in the Google Developer's console it won't pair. If you haven't done it, go to 

https://console.home.google.com/projects

Create a new project - I think the only important values are the vendor ID of 0xfff1 and the product id of 0x8000. We've been using Acme Light Bulb for the product and Light for the type although others work.

Otherwise for debug, you can run "add logcat | grep Matter" if you have the Android Debugger installed on your computer. That will show all the Android debug messages flying by

**Reply by javierbarellano4v** [reply]
@JimBuzbee Wow! that worked! got it paired/onboarded in Google Home. Thanks!

**Reply by javierbarellano4v** [reply]
I missed that step completely - project creation!

**Reply by javierbarellano4v** [reply]
To update on Alexa use-case: shifted to a new environment. Now able to pair with node-matter. Very strange wasn't getting scanning step.

Additionally, able get new pairing code from Alexa app and then paired a different controller, chip-tool. So now can control 'light' via Alexa or chip-tool with node-matter paired with both.
